plugins/woocommerce-admin/client/products/hooks/use-product-auto-save.ts

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import { debounce, pick } from 'lodash';
+import { Product } from '@woocommerce/data';
+import { useDebounce } from '@wordpress/compose';
+import { useEffect, useMemo, useState } from '@wordpress/element';
+import { useFormContext } from '@woocommerce/components';
+
+/**
+ * Internal dependencies
+ */
+import { useProductHelper } from '../use-product-helper';
+
+export const useProductAutoSave = ( dependencies: ( keyof Product )[] ) => {
+	const [ isAutoSaving, setIsAutoSaving ] = useState( false );
+	const { isValidForm, setValue, values } = useFormContext< Product >();
+	const { createOrUpdateProductWithStatus } = useProductHelper();
+
+	const dependencyValues = dependencies.reduce(
+		( acc: unknown[], key: keyof Product ) => {
+			acc.push( values[ key ] );
+			return acc;
+		},
+		[]
+	);
+
+	const handleProductUpdated = ( updatedProduct: Product ) => {
+		if ( ! updatedProduct || ! updatedProduct.id ) {
+			return;
+		}
+
+		setValue( 'id', updatedProduct.id );
+		setIsAutoSaving( false );
+	};
+
+	const saveProduct = () => {
+		setIsAutoSaving( true );
+		createOrUpdateProductWithStatus(
+			values.id || null,
+			pick( values, dependencies ),
```

Comment:
I used a string identifying the property key (`name`) opposed to the actual value (`values.name`) so that we can only update those values and not the entire form.  I think this is what we want, but this may change in the future depending on how we use this.